core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -2914,6 +2928,111 @@ public void testSequenceIDs() throws Exception {
+        final long globalCheckpoint =
+            localCheckpoint == SequenceNumbersService.NO_OPS_PERFORMED || rarely()
+                ? SequenceNumbersService.UNASSIGNED_SEQ_NO
+                : Math.toIntExact(localCheckpoint);
+        Engine initialEngine = null;
+        try {
+            initialEngine = createEngine(
+                defaultSettings,
+                store,
+                primaryTranslogDir,
+                newMergePolicy(),
+                null,
+                () -> new SequenceNumbersService(shardId, defaultSettings, maxSeqNo, localCheckpoint, globalCheckpoint) {
+                    @Override
+                    public long generateSeqNo() {
+                        if (rarely()) {
+                            // force skipping a sequence number
+                            super.generateSeqNo();
+                        }
+                        return super.generateSeqNo();
+                    }
+                });
+            initialEngine.flush(true, true);
+            final int docs = randomIntBetween(1, 32);
+            for (int i = 0; i < docs; i++) {
+                final String id = Integer.toString(i);
+                final ParsedDocument doc = testParsedDocument(id, id, "test", null, testDocumentWithTextField(), SOURCE, null);
+                initialEngine.index(new Engine.Index(newUid(id), doc));
+            }
+        } finally {
+            IOUtils.close(initialEngine);
+        }
+
+        try (final Engine recoveringEngine =
+                 new InternalEngine(copy(initialEngine.config(), EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG))) {
+            // there might be gaps in the translog; the best that we can say is that the local checkpoint is at least as far the max
+            // sequence number
+            // TODO: tighten this assertion after gaps in the translog are addressed
+            assertThat(recoveringEngine.seqNoService().getLocalCheckpoint(), greaterThanOrEqualTo(maxSeqNo));
+        }
+    }
+
+    public void testOutOfOrderSequenceNumbersWithVersionConflict() {
+        final List<Engine.Operation> operations = new ArrayList<>();
+        final int numberOfOperations = randomIntBetween(16, 32);
+        final Term uid = newUid("1");
+        final Document document = testDocumentWithTextField();
+        document.add(new Field(SourceFieldMapper.NAME, BytesReference.toBytes(B_1), SourceFieldMapper.Defaults.FIELD_TYPE));
+        final ParsedDocument doc = testParsedDocument("1", "1", "test", null, document, B_1, null);
+        for (int i = 0; i < numberOfOperations; i++) {
+            if (randomBoolean()) {
+                operations.add(indexOperation(uid, doc, i, i));
+            } else {
+                operations.add(deleteOperation("test", "1", uid, i, i));
+            }
+        }
+
+        final boolean exists = operations.get(operations.size() - 1) instanceof Engine.Index;
```

Comment:
This represents whether or not the last operation before shuffling, and hence the operation with the highest version, is an index or delete operation. I think that this is correct?